### ISSUE TITLE
MigrateFileTask leaves behind original files if they were renamed

### ISSUE THREAD

**Issue by gheggie** [silverstripe/silverstripe-assets#216]
The task creates a new file if it was renamed during the name generation process, and leaves a copy of the original file in the filesystem (at the same hash directory location as the new file). 

It seems like in this case the original is essentially an orphan file; if the "new" file is moved to the protected location with `protectFile()` the original is left behind and remains publicly accessible.

**Comment by gheggie** [comment]
On a SS4.3.2 version install, this issue is caused by File objects with consecutive dashes in the file name being renamed during `File::write()`. A file named `My---Document.txt` will be renamed to `My-Document.txt`. This same behaviour can also be confirmed by uploading a file in the Files interface of the CMS with consecutive dashes in the name.

I don't think there's any (logical) reason for consecutive dashes in a file name to be reduced to a single dash though.

**Comment by maxime-rainville** [comment]
Yeah we do have some problems with the file migration task leaving some old files behind. Old variants being the most egregious one right now. 

I don't see us changing the file naming rules. I'll have to have another look at the file migration script as part of https://github.com/silverstripe/silverstripe-versioned/issues/177. I'll see if I can update the migration task so it better handles the renaming.

**Comment by maxime-rainville** [comment]
Yeah we do have some problems with the file migration task leaving some old files behind. Old variants being the most egregious one right now. 

I don't see us changing the file naming rules. I'll have to have another look at the file migration script as part of https://github.com/silverstripe/silverstripe-versioned/issues/177. I'll see if I can update the migration task so it better handles the renaming.

**Comment by gheggie** [comment]
The only thing I'd like to add in response to

> I don't see us changing the file naming rules

is that the rules don't really make sense. Why should the files name `file-name.jpg`, `file--name.jpg` and `file---name.jpg` not be considered as three different/distinct files and have their names preserved accordingly? SilverStripe will enforce the difference by normalizing the name and then adding suffixes - but it makes no sense as to *why* this should happen?

**Comment by maxime-rainville** [comment]
I haven't look at this in detail, but my guess is we probably have some internal logic that assumes some pre-defined pattern in the file names. Untangling this might be a bit of mission. 

For example, if we allowed people include two consecutive underscores in a folder name, this would completely break our URL resolution scheme for thumbnails and other resized images.

**Comment by chillu** [comment]
@maxime-rainville Can you check what we should do with this issue please?

**Comment by maxime-rainville** [comment]
Some of the latest changes we've been doing. Will affect this. I didn't specifically check what happens with multi-dashes names. I'll investigate that some more.

@gheggie We ended up changing the way we generate file name in SS44, which would allow you to define a custom format ... in theory. That API is marked as experimental, so we don't support it officially.
